cpp/src/ffi/reader_writer_jni.cpp

```diff
@@ -0,0 +1,512 @@
+    const char* str = env->GetStringUTFChars(jstr, nullptr);
+    strings[i] = strdup(str);
+    env->ReleaseStringUTFChars(jstr, str);
+    env->DeleteLocalRef(jstr);
+  }
+
+  return strings;
+}
+
+void FreeStringArray(JNIEnv* env, const char** strings, size_t count) {
+  if (strings != nullptr) {
+    for (size_t i = 0; i < count; ++i) {
+      free(const_cast<char*>(strings[i]));
+    }
+    free(strings);
+  }
+}
+
+// ==================== JNI Properties Implementation ====================
+
+JNIEXPORT void JNICALL
+Java_io_milvus_storage_MilvusStorageProperties_createProperties(JNIEnv* env, jobject obj, jobject java_map, jlong properties_ptr) {
+  try {
+    Properties* properties = reinterpret_cast<Properties*>(properties_ptr);
+
+    jclass map_class = env->GetObjectClass(java_map);
+    jmethodID entry_set_method = env->GetMethodID(map_class, "entrySet", "()Ljava/util/Set;");
+    jobject entry_set = env->CallObjectMethod(java_map, entry_set_method);
+
+    jclass set_class = env->GetObjectClass(entry_set);
+    jmethodID to_array_method = env->GetMethodID(set_class, "toArray", "()[Ljava/lang/Object;");
+    jobjectArray entries = static_cast<jobjectArray>(env->CallObjectMethod(entry_set, to_array_method));
+
+    jsize num_entries = env->GetArrayLength(entries);
+
+    std::vector<const char*> keys, values;
+    std::vector<std::string> key_storage, value_storage;
+
+    for (jsize i = 0; i < num_entries; ++i) {
+      jobject entry = env->GetObjectArrayElement(entries, i);
+      jclass entry_class = env->GetObjectClass(entry);
+
+      jmethodID get_key_method = env->GetMethodID(entry_class, "getKey", "()Ljava/lang/Object;");
+      jmethodID get_value_method = env->GetMethodID(entry_class, "getValue", "()Ljava/lang/Object;");
+
+      jstring key_jstr = static_cast<jstring>(env->CallObjectMethod(entry, get_key_method));
+      jstring value_jstr = static_cast<jstring>(env->CallObjectMethod(entry, get_value_method));
+
+      const char* key_cstr = env->GetStringUTFChars(key_jstr, nullptr);
+      const char* value_cstr = env->GetStringUTFChars(value_jstr, nullptr);
+
+      key_storage.emplace_back(key_cstr);
+      value_storage.emplace_back(value_cstr);
+
+      env->ReleaseStringUTFChars(key_jstr, key_cstr);
+      env->ReleaseStringUTFChars(value_jstr, value_cstr);
+      env->DeleteLocalRef(key_jstr);
+      env->DeleteLocalRef(value_jstr);
+      env->DeleteLocalRef(entry);
+    }
+
+    for (size_t i = 0; i < key_storage.size(); ++i) {
+      keys.push_back(key_storage[i].c_str());
+      values.push_back(value_storage[i].c_str());
+    }
+
+    FFIResult result = properties_create(keys.data(), values.data(), keys.size(), properties);
+    if (!IsSuccess(&result)) {
+      ThrowJavaExceptionFromFFIResult(env, &result);
+      FreeFFIResult(&result);
+    }
+
+    env->DeleteLocalRef(entries);
+    env->DeleteLocalRef(entry_set);
+  } catch (...) {
```

Comment:
i don't get, L149 throw a java exception or a C++ exception, will the C++ `try...catch` catch the exception ?

Reply:
also will `DeleteLocalRef` throw a exception? if `DeleteLocalRef` throw a exception , in L155 catch the exception, then some of memory will be leak?